onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -200,104 +206,131 @@ Status TreeEnsembleCommon<InputType, ThresholdType, OutputType>::Init(
+                                      : nodes_values_as_tensor[i];
+
+    /* hitrates is not used for inference, they are ignored.
     if (nodes_hitrates_as_tensor.empty()) {
       node.hitrates = static_cast<ThresholdType>(i < nodes_hitrates.size() ? nodes_hitrates[i] : -1);
     } else {
       node.hitrates = i < nodes_hitrates_as_tensor.size() ? nodes_hitrates_as_tensor[i] : -1;
+    } */
+
+    node.flags |= static_cast<uint8_t>(cmodes[i]);
+    node.truenode_inc_or_first_weight = 0;  // nodes_truenodeids[i] if not a leaf
+    node.falsenode_inc_or_n_weights = 0;    // nodes_falsenodeids[i] if not a leaf
+
+    if (i < static_cast<size_t>(nodes_missing_value_tracks_true.size()) && nodes_missing_value_tracks_true[i] == 1) {
+      node.flags |= static_cast<uint8_t>(MissingTrack::kTrue);
     }
-    node.mode = cmodes[i];
-    node.is_not_leaf = node.mode != NODE_MODE::LEAF;
-    node.truenode = nullptr;   // nodes_truenodeids[i];
-    node.falsenode = nullptr;  // nodes_falsenodeids[i];
-    node.missing_tracks = i < static_cast<size_t>(nodes_missing_value_tracks_true.size())
-                              ? (nodes_missing_value_tracks_true[i] == 1
-                                     ? MissingTrack::kTrue
-                                     : MissingTrack::kFalse)
-                              : MissingTrack::kNone;
-    node.is_missing_track_true = node.missing_tracks == MissingTrack::kTrue;
-    if (idi.find(node.id) != idi.end()) {
-      ORT_THROW("Node ", node.id.node_id, " in tree ", node.id.tree_id, " is already there.");
+    if (idi.find(node_tree_id) != idi.end()) {
+      ORT_THROW("Node ", node_tree_id.node_id, " in tree ", node_tree_id.tree_id, " is already there.");
     }
-    idi.insert(std::pair<TreeNodeElementId, TreeNodeElement<ThresholdType>*>(node.id, &node));
+    idi.insert(std::pair<TreeNodeElementId, uint32_t>(node_tree_id, i));
   }
 
   TreeNodeElementId coor;
+  i = 0;
   for (auto it = nodes_.begin(); it != nodes_.end(); ++it, ++i) {
-    if (!it->is_not_leaf)
+    if (!it->is_not_leaf())
       continue;
-    i = std::distance(nodes_.begin(), it);
-    coor.tree_id = it->id.tree_id;
+
+    TreeNodeElementId& node_tree_id = node_tree_ids[i];
+    coor.tree_id = node_tree_id.tree_id;
     coor.node_id = static_cast<int>(nodes_truenodeids[i]);
 
     auto found = idi.find(coor);
     if (found == idi.end()) {
       ORT_THROW("Unable to find node ", coor.tree_id, "-", coor.node_id, " (truenode).");
     }
-    if (coor.node_id >= 0 && coor.node_id < n_nodes_) {
-      it->truenode = found->second;
-      if ((it->truenode->id.tree_id != it->id.tree_id) ||
-          (it->truenode->id.node_id == it->id.node_id)) {
-        ORT_THROW("One falsenode is pointing either to itself, either to another tree.");
-      }
-    } else
-      it->truenode = nullptr;
+    truenode_ids[i] = (coor.node_id >= 0 && coor.node_id < n_nodes_) ? found->second : 0;
```

Comment:
> truenode_ids[i] = (coor.node_id >= 0 && coor.node_id < n_nodes_) ? found->second : 0; [](http://example.com/codeflow?start=3&length=86)

Consider breaking up computations and insert for better readability possibly with some comments.
Suggesting use of insert_or_assign() for better readability